From 7d8970e53a062ad7de6711deaf0b2998db739044 Mon Sep 17 00:00:00 2001
From: Dhina17 <dhinalogu@gmail.com>
Date: Mon, 11 Mar 2024 22:21:23 +0000
Subject: [PATCH 16/35] services: Workaround for our face unlock initialization
 [1/2]

AuthService registers the authenticators only once and
then onAllAuthenticatorsRegistered() callbacks are invoked.
The callbacks are removed as soon as it invoked.

Our LMO face unlock authenticator isn't being registered via AuthService.
When we register it, the onAllAuthenticatorsRegistered() callbacks aren't available
at that time so our face unlock isn't being initialized properly
and end up with bugs like "Face unlock isn't working after reboot".

As a workaround, we don't invoke callbacks if it's first time (from AuthService).
Instead we register our face unlock here on first time to make sure
callbacks are being called only after our face unlock authenticator registration.

If properties aren't empty, it's from hardware face sensor.
so allow to invoke callbacks regardless of the first time.

Change-Id: I26238a35f09d9cb398355d9467d44cd24b0e2525
---
 .../sensors/BiometricServiceRegistry.java     |  41 +++++-
 .../biometrics/sensors/face/FaceService.java  |   2 +-
 .../sensors/face/FaceServiceRegistry.java     |   5 +-
 .../fingerprint/FingerprintService.java       |   3 +-
 .../FingerprintServiceRegistry.java           |   5 +-
 .../libremobileos/FaceUnlockHelper.java       | 131 ++++++++++++++++++
 6 files changed, 177 insertions(+), 10 deletions(-)
 create mode 100644 services/core/java/com/android/server/libremobileos/FaceUnlockHelper.java

diff --git a/services/core/java/com/android/server/biometrics/sensors/BiometricServiceRegistry.java b/services/core/java/com/android/server/biometrics/sensors/BiometricServiceRegistry.java
index 5a493e4dded8..f7dcc10f43ce 100644
--- a/services/core/java/com/android/server/biometrics/sensors/BiometricServiceRegistry.java
+++ b/services/core/java/com/android/server/biometrics/sensors/BiometricServiceRegistry.java
@@ -19,6 +19,7 @@ package com.android.server.biometrics.sensors;
 
 import android.annotation.NonNull;
 import android.annotation.Nullable;
+import android.content.Context;
 import android.hardware.biometrics.IBiometricAuthenticator;
 import android.hardware.biometrics.IBiometricService;
 import android.hardware.biometrics.SensorPropertiesInternal;
@@ -32,6 +33,8 @@ import android.util.Slog;
 
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.server.ServiceThread;
+import com.android.server.biometrics.sensors.face.FaceServiceRegistry;
+import com.android.server.libremobileos.FaceUnlockHelper;
 
 import java.util.ArrayList;
 import java.util.Collections;
@@ -63,8 +66,14 @@ public abstract class BiometricServiceRegistry<T extends BiometricServiceProvide
     @NonNull
     private final RemoteCallbackList<C> mRegisteredCallbacks = new RemoteCallbackList<>();
 
-    public BiometricServiceRegistry(@NonNull Supplier<IBiometricService> biometricSupplier) {
+    @NonNull
+    private final FaceUnlockHelper mFaceUnlockHelper;
+    private volatile boolean mFirstTime = true;
+
+    public BiometricServiceRegistry(@NonNull Supplier<IBiometricService> biometricSupplier,
+            Context context) {
         mBiometricServiceSupplier = biometricSupplier;
+        mFaceUnlockHelper = new FaceUnlockHelper(context);
     }
 
     /**
@@ -134,9 +143,33 @@ public abstract class BiometricServiceRegistry<T extends BiometricServiceProvide
 
     private synchronized void finishRegistration(
             @NonNull List<T> providers, @NonNull List<P> allProps) {
-        mServiceProviders = Collections.unmodifiableList(providers);
-        mAllProps = Collections.unmodifiableList(allProps);
-        broadcastAllAuthenticatorsRegistered();
+        // AuthService registers the authenticators only once and
+        // then onAllAuthenticatorsRegistered() callbacks are invoked.
+        // The callbacks are removed as soon as it invoked.
+        // Our LMO face unlock authenticator isn't being registered via AuthService.
+        // When we register it, the onAllAuthenticatorsRegistered() callbacks aren't available
+        // at that time so our face unlock won't be initialized properly
+        // and end up with bugs like "Face unlock isn't working after reboot".
+        // As a workaround, we don't invoke callbacks if it's first time (from AuthService).
+        // Instead we register our face unlock here on first time to make sure
+        // callbacks are being called only after our face authenticator registration.
+        // If properties aren't empty, it's from hardware face sensor.
+        // so allow to invoke callbacks regardless of the first time.
+        if (this instanceof FaceServiceRegistry) {
+            if (!allProps.isEmpty() || !mFirstTime) {
+                mServiceProviders = Collections.unmodifiableList(providers);
+                mAllProps = Collections.unmodifiableList(allProps);
+                broadcastAllAuthenticatorsRegistered();
+            } else {
+                mFirstTime = false;
+                mFaceUnlockHelper.registerLMOFaceAuthenticator();
+            }
+        } else {
+            // Allow as like before for authenticators other than face.
+            mServiceProviders = Collections.unmodifiableList(providers);
+            mAllProps = Collections.unmodifiableList(allProps);
+            broadcastAllAuthenticatorsRegistered();
+        }
     }
 
     /**
diff --git a/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java b/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
index 8c988729a1ae..a455d6db1181 100644
--- a/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
+++ b/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
@@ -769,7 +769,7 @@ public class FaceService extends SystemService {
         mLockPatternUtils = new LockPatternUtils(context);
         mBiometricStateCallback = new BiometricStateCallback<>(UserManager.get(context));
         mAuthenticationStateListeners = new AuthenticationStateListeners();
-        mRegistry = new FaceServiceRegistry(mServiceWrapper, biometricServiceSupplier);
+        mRegistry = new FaceServiceRegistry(mServiceWrapper, biometricServiceSupplier, context);
         mRegistry.addAllRegisteredCallback(new IFaceAuthenticatorsRegisteredCallback.Stub() {
             @Override
             public void onAllAuthenticatorsRegistered(List<FaceSensorPropertiesInternal> sensors) {
diff --git a/services/core/java/com/android/server/biometrics/sensors/face/FaceServiceRegistry.java b/services/core/java/com/android/server/biometrics/sensors/face/FaceServiceRegistry.java
index 0f0a81d24473..f11a6def1570 100644
--- a/services/core/java/com/android/server/biometrics/sensors/face/FaceServiceRegistry.java
+++ b/services/core/java/com/android/server/biometrics/sensors/face/FaceServiceRegistry.java
@@ -20,6 +20,7 @@ import static android.hardware.biometrics.BiometricAuthenticator.TYPE_FACE;
 
 import android.annotation.NonNull;
 import android.annotation.Nullable;
+import android.content.Context;
 import android.hardware.biometrics.BiometricManager;
 import android.hardware.biometrics.IBiometricService;
 import android.hardware.face.FaceSensorPropertiesInternal;
@@ -45,8 +46,8 @@ public class FaceServiceRegistry extends BiometricServiceRegistry<ServiceProvide
 
     /** Creates a new registry tied to the given service. */
     public FaceServiceRegistry(@NonNull IFaceService service,
-            @Nullable Supplier<IBiometricService> biometricSupplier) {
-        super(biometricSupplier);
+            @Nullable Supplier<IBiometricService> biometricSupplier, Context context) {
+        super(biometricSupplier, context);
         mService = service;
     }
 
diff --git a/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintService.java b/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintService.java
index 2f6ba0b852ee..565a159242ea 100644
--- a/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintService.java
+++ b/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintService.java
@@ -1068,7 +1068,8 @@ public class FingerprintService extends SystemService {
                                         resetLockoutRequiresHardwareAuthToken);
 
         mHandler = new Handler(Looper.getMainLooper());
-        mRegistry = new FingerprintServiceRegistry(mServiceWrapper, biometricServiceSupplier);
+        mRegistry =
+            new FingerprintServiceRegistry(mServiceWrapper, biometricServiceSupplier, context);
         mRegistry.addAllRegisteredCallback(new IFingerprintAuthenticatorsRegisteredCallback.Stub() {
             @Override
             public void onAllAuthenticatorsRegistered(
diff --git a/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintServiceRegistry.java b/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintServiceRegistry.java
index 33810b764f23..d86432d8d1a2 100644
--- a/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintServiceRegistry.java
+++ b/services/core/java/com/android/server/biometrics/sensors/fingerprint/FingerprintServiceRegistry.java
@@ -20,6 +20,7 @@ import static android.hardware.biometrics.BiometricAuthenticator.TYPE_FINGERPRIN
 
 import android.annotation.NonNull;
 import android.annotation.Nullable;
+import android.content.Context;
 import android.hardware.biometrics.BiometricManager;
 import android.hardware.biometrics.IBiometricService;
 import android.hardware.fingerprint.FingerprintSensorPropertiesInternal;
@@ -45,8 +46,8 @@ public class FingerprintServiceRegistry extends BiometricServiceRegistry<Service
 
     /** Creates a new registry tied to the given service. */
     public FingerprintServiceRegistry(@NonNull IFingerprintService service,
-            @Nullable Supplier<IBiometricService> biometricSupplier) {
-        super(biometricSupplier);
+            @Nullable Supplier<IBiometricService> biometricSupplier, Context context) {
+        super(biometricSupplier, context);
         mService = service;
     }
 
diff --git a/services/core/java/com/android/server/libremobileos/FaceUnlockHelper.java b/services/core/java/com/android/server/libremobileos/FaceUnlockHelper.java
new file mode 100644
index 000000000000..d9ae8d454734
--- /dev/null
+++ b/services/core/java/com/android/server/libremobileos/FaceUnlockHelper.java
@@ -0,0 +1,131 @@
+/**
+ * Copyright (C) 2024 The LibreMobileOS Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.server.libremobileos;
+
+import static android.hardware.biometrics.BiometricAuthenticator.TYPE_FACE;
+import static android.hardware.biometrics.BiometricManager.Authenticators;
+
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.hardware.biometrics.BiometricManager;
+import android.hardware.biometrics.IBiometricService;
+import android.hardware.face.IFaceService;
+import android.hardware.face.FaceSensorConfigurations;
+import android.hardware.face.FaceSensorProperties;
+import android.hardware.face.FaceSensorPropertiesInternal;
+import android.hardware.fingerprint.FingerprintManager;
+import android.hardware.fingerprint.FingerprintSensorPropertiesInternal;
+import android.hardware.fingerprint.IFingerprintAuthenticatorsRegisteredCallback;
+import android.os.RemoteException;
+import android.os.ServiceManager;
+import android.util.Slog;
+
+import com.android.internal.R;
+import com.android.server.biometrics.BiometricHandlerProvider;
+import com.android.server.biometrics.Utils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FaceUnlockHelper {
+
+    private static final String TAG = "FaceUnlockHelper";
+
+    private final Context mContext;
+
+    public FaceUnlockHelper(Context context) {
+        mContext = context;
+    }
+
+    public void registerLMOFaceAuthenticator() {
+        final PackageManager pm = mContext.getPackageManager();
+        boolean supportsFace = pm.hasSystemFeature(PackageManager.FEATURE_FACE);
+        // Ignore if device don't support it
+        if (!supportsFace) return;
+
+        // We have to wait until fingerprint authenticators are registered
+        // so we can get the fingerprint sensor IDs to avoid duplication
+        // with our randomly generated face sensor ID.
+        FingerprintManager fingerprintManager =
+                mContext.getSystemService(FingerprintManager.class);
+        if (fingerprintManager != null) {
+            fingerprintManager.addAuthenticatorsRegisteredCallback(
+                    new IFingerprintAuthenticatorsRegisteredCallback.Stub() {
+                @Override
+                public void onAllAuthenticatorsRegistered(
+                        List<FingerprintSensorPropertiesInternal> fingerprintSensors) {
+                    initializeIds();
+                }
+            });
+        } else {
+            initializeIds();
+        }
+    }
+
+    private void initializeIds() {
+        try {
+            IBiometricService biometricService = IBiometricService.Stub.asInterface(
+                    ServiceManager.getService(Context.BIOMETRIC_SERVICE));
+            if (biometricService == null) {
+                Slog.e(TAG, "Face feature exists, but IBiometricService is null.");
+                return;
+            }
+            IFaceService faceService = IFaceService.Stub.asInterface(
+                    ServiceManager.getService(Context.FACE_SERVICE));
+            if (faceService == null) {
+                Slog.e(TAG, "Face feature exists, but FaceService is null.");
+                return;
+            }
+            List<FaceSensorPropertiesInternal> faceSensors =
+                    faceService.getSensorPropertiesInternal(mContext.getOpPackageName());
+            if (faceSensors.size() == 0) {
+                Slog.i(TAG, "Using software face sensor.");
+                int newId = 0;
+                // IDs may come from HALs and be non-linear, ensure we really have unique ID,
+                // because if ID is duplicated, we crash system server!
+                boolean foundDuplicate = false;
+                do {
+                    if (foundDuplicate) {
+                        newId++;
+                    }
+                    foundDuplicate = biometricService.getCurrentStrength(newId)
+                            != Authenticators.EMPTY_SET;
+                } while (foundDuplicate);
+                final int faceId = newId;
+                BiometricHandlerProvider.getInstance().getFaceHandler().post(() -> {
+                    try {
+                        faceService.registerAuthenticators(getFaceSensorConfigurations(mContext, faceId));
+                    } catch (RemoteException e) {
+                        Slog.e(TAG, "RemoteException when registering face authenticators", e);
+                    }
+                });
+            } else {
+                Slog.i(TAG, "Using hardware face sensor.");
+            }
+        } catch (RemoteException e) {
+            Slog.e(TAG, "RemoteException when loading face configuration", e);
+        }
+    }
+
+    private FaceSensorConfigurations getFaceSensorConfigurations(Context context, int sensorId) {
+        final FaceSensorConfigurations faceSensorConfigurations =
+                    new FaceSensorConfigurations(true /*  resetLockoutRequiresChallenge */);
+        String sensorConfig = sensorId + ":" + TYPE_FACE + ":" + Authenticators.BIOMETRIC_STRONG;
+        String[] hidlConfigStrings = { sensorConfig };
+        faceSensorConfigurations.addHidlConfigs(hidlConfigStrings, context);
+        return faceSensorConfigurations;
+    }
+}
-- 
2.49.0

